### DISCUSSION TITLE
Controlling a matrix of LEDs

### DISCUSSION THREAD

**Discussion by JovoSK** [tttapa/Control-Surface#428] [Q&A]
Hi folks,

I'm attempting to gain individual control over a matrix of LEDs, such that they would be individually toggle-able via MIDI message. I'm pretty (read: very) new to this. Matrix is 10x5 and on PCB, which makes it awkward to try to retrofit a MAX7219 in.

My thought is to array the LEDs, but I'm unsure of how to proceed.

Any advice would be greatly appreciated.

Thanks!

**EDIT**

After some further research, I found this: 

```cpp
/*
 * _8x8 LED matrix
 *
 * (c) 2009 BlushingBoy.net
 */

#include "TimerOne.h"

byte rows[8] = {9, 14, 8, 12, 1, 7, 2, 5};
byte cols[8] = {13, 3, 4, 10, 6, 11, 15, 16};
byte pins[16] = {5, 4, 3, 2, 14, 15, 16, 17, 13, 12, 11, 10, 9, 8, 7, 6};
byte screen[8] = {0, 0, 0, 0, 0, 0, 0, 0};
volatile byte screenRow = 0;
volatile byte screenCol = 0;

void setup() {
  Timer1.initialize(100);
  for (int i = 2; i <= 17; i++)
    pinMode(i, OUTPUT);
  Timer1.attachInterrupt(doubleBuffer);
}

// interrupt routine
void doubleBuffer() {
  // reset the previous iteration
  digitalWrite(translatePin(rows[screenRow]), HIGH); // set previous off
  digitalWrite(translatePin(cols[screenCol]), LOW);  // set previous off
  // go to the next iteration...
  // go to the next screenCol, wrap if necessary
  screenCol++;
  if (screenCol >= 8) {
    screenCol = 0;
    // when screenCol wraps, go to the next screenRow, wrap if necessary
    screenRow++;
    if (screenRow >= 8) {
      screenRow = 0;
    }
  }
  // set this iteration
  if((screen[screenRow]>>screenCol)&B1 == B1) {
    digitalWrite(translatePin(rows[screenRow]), LOW);  // set this on
    digitalWrite(translatePin(cols[screenCol]), HIGH); // set this on
  } else {
    digitalWrite(translatePin(rows[screenRow]), HIGH); // set this off
    digitalWrite(translatePin(cols[screenCol]), LOW);  // set this off
  }
}

byte translatePin(byte original) {
  return pins[original - 1];
}

void allOFF() {
  for (int i = 0; i < 8; i++)
    screen[i]=0;
}

void on(byte row, byte column) {
    screen[column-1] |= (B1<<(row-1));
}

void off(byte row, byte column) {
    screen[column-1] &= ~(B1<<(row-1));
}
```

Source: https://playground.arduino.cc/Main/DirectDriveLEDMatrix/?fbclid=IwAR3BCcRQW4homrzA1M3msrANngjVmotSv2Xv7VfaTd2U4r5nTxAFkYVlZpc

This seems to do what I want in regards to interacting with the matrix. The follow up then is, how do I integrate the CS library in, such that when the Arduino receives a note from the PC, it fires the appropriate LEDs?

**Comment by tttapa** [comment]
I'd recommend using the `new-input` branch of this library.

Then you can implement your own `NoteLED` class that matches incoming MIDI note messages and turns on/off an LED accordingly:
https://github.com/tttapa/Control-Surface/blob/a061a3a0f021059bfd53c89c928271476714be7b/src/MIDI_Inputs/LEDs/NoteCCKPLED.hpp#L10-L56

Replace the calls to `digitalWrite` on lines 35 and 46 by calls to the `on()` or `off()` function in your sketch to turn on/off the correct LED. You can probably leave the `begin()` method empty.
